guy/my_ps/proc_info.c

```diff
@@ -0,0 +1,131 @@
+//
+// Created by root on 4/24/20.
+//
+
+#include "proc_info.h"
+
+
+/*
+ * Copies the file's content in a buffer. Built to work with the proc fs.
+ * path [IN] -> The file's path.
+ * file_contents [OUT] -> buffer to which we copy the file's contents.
+ * dereference_symlinks [IN] -> If true, read the file the symlink point to rather than the link file itself.
+* RETURN -> Number of bytes read.
+ */
+size_t read_proc_fs_file(const char *path, char *file_contents, bool dereference_symlinks) {
+    size_t bytes_read = 0;
+    struct stat file_stat;
+
+    if (fstatat(0, path, &file_stat, AT_SYMLINK_NOFOLLOW)) {
+        if (DEBUG) {
+            printf("Couldn't stat \"%s\". Check it exists and has the right permissions set.\n", path);
+        }
+
+        file_contents = NULL;
+        return 0;
+    }
+    // If told not to dereference links, check if the file is a link and if it is return its target.
+    if (!dereference_symlinks && S_ISLNK(file_stat.st_mode)) {
+        bytes_read = readlink(path, file_contents, PATH_MAX - 1);
+        file_contents[bytes_read] = '\0';
+        return bytes_read;
+    }
+
+    FILE *fd = fopen(path, "r");
+    if (NULL == fd) {
+        if (DEBUG) {
+            printf("An Error has occurred while opening \"%s\".\n", path);
+        }
+    } else {
+        // Since it turns out we cant use lseek or stat to find out file size, we will to read
+        //bytes_read = fread((void*)file_contents, MAX_PROC_FS_FILE_SIZE, 1, fd);
+        for (bytes_read = 0;
+             bytes_read < MAX_PROC_FS_FILE_SIZE && EOF != (file_contents[bytes_read] = fgetc(fd)); bytes_read++);
```

Comment:
what will happen to file_contents if the file's content is larger than MAX_PROC_FS_FILE_SIZE?

Reply:
That's a good question and I've spent much time trying to come up with a good solution to the obvious problem that might arise.
Problem is that when you stat a file in the proc fs it always returns st_size = 0 and fseek doesn't seem to help either. 
I could open the file twice - once to count it's size and the other to read it properly once iv'e allocated enough space but it would be extremely wasteful (not to mention that it won't necessarily work as the proc fs might generate different data every time we open the file)

Reply:
I know :) and it really would have been pointless.
my point is that if the file's content is larger than MAX_PROC_FS_FILE_SIZE, file_contents won't be a proper string as it won't end in \0.

Reply:
Well, that's a fair point. Fixed